cmd/zstream/zstream_recompress.c

```diff
@@ -155,6 +157,8 @@ zstream_do_recompress(int argc, char *argv[])
 		case DRR_BEGIN:
 		{
 			ZIO_SET_CHECKSUM(&stream_cksum, 0, 0, 0, 0);
+			VERIFY3B(seen, ==, B_TRUE);
+			begin--;
```

Comment:
Did you mean
```suggestion
			VERIFY0(begin++);
			seen = B_TRUE;
```
?

Reply:
No. In a replication stream, there is 1 END record without a corresponding BEGIN record to indicate the true end. This hack handles that. If it were not for that, we would have been able to do what you suggested, minus the seen variable, which would be unnecessary. An earlier version of this patch did that, but it broke replication streams, which the buildbot caught.

Reply:
I have trouble understanding how this could work fine when `seen` is never set to `B_TRUE` and `begin` is only decremented, starting from 0. Yet
```
VERIFY3B(seen, ==, B_TRUE);
```
and
```
VERIFY3S(begin, ==, 1);
```
don't throw.

Reply:
`seen` is set to `B_TRUE` in `DRR_BEGIN`. We don't want to allow us to process a `DRR_END` record without having seen a `DRR_BEGIN` record, so this works. Also, `begin` is incremented in `DRR_BEGIN` via `VERIFY0(begin++);`.

For the other record types, we want to be inside a DRR_BEGIN and DDR_END, so `VERIFY3S(begin, ==, 1);` verifies that.

Reply:
What might be confusing is that the less than perfect constraints allow you process a stream with an infinite number of DDR_END records at its tail (and if you manage to wraparound, then process other record types too). That is an imperfection that I was willing to tolerate to handle the case of the unmatched `DDR_END` record in replication streams.

Reply:
I get the idea for the other 2 files. I fail to see `seen` being set in `cmd/zstream/zstream_recompress.c` to `B_TRUE` at all.
Does it mean that we don't execute those paths at all or possibly we don't have enough coverage in ZTS?

Reply:
Oh, I see it now. You are right that I goofed on this file. I will repush a fix.

That said, you are correct that we do not have enough coverage in the ZTS. It would have caught this mistake in other files, but it apparently leaks a test to catch it in this one.

Reply:
No worries. The solution is indeed a bit tricky.

Reply:
A correction has been pushed. Thanks again.